coherence/v1client.go

```diff
@@ -0,0 +1,1371 @@
+	req, err := m.newPageOfEntriesRequest(cache, cookie)
+	if err != nil {
+		return nil, err
+	}
+
+	requestType, err := m.submitRequest(req, pb1.NamedCacheRequestType_PageOfEntries)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.getStreamingResponseKeyAndValue(ctx, requestType, req.Id, true)
+}
+
+func (m *streamManagerV1) keyPage(ctx context.Context, cache string, cookie []byte) (<-chan BinaryKey, error) {
+	req, err := m.newPageOfKeysRequest(cache, cookie)
+	if err != nil {
+		return nil, err
+	}
+
+	requestType, err := m.submitRequest(req, pb1.NamedCacheRequestType_PageOfKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.getStreamingResponseKey(ctx, requestType, req.Id, true)
+}
+
+func (m *streamManagerV1) invoke(ctx context.Context, cache string, agent []byte, keyOrFilter *pb1.KeysOrFilter) (<-chan BinaryKeyAndValue, error) {
+	req, err := m.newInvokeRequest(cache, agent, keyOrFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	requestType, err := m.submitRequest(req, pb1.NamedCacheRequestType_Invoke)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.getStreamingResponseKeyAndValue(ctx, requestType, req.Id)
+}
+
+func (m *streamManagerV1) entrySetFilter(ctx context.Context, cache string, filter []byte) (<-chan BinaryKeyAndValue, error) {
+	req, err := m.newEntrySetRequest(cache, filter)
+	if err != nil {
+		return nil, err
+	}
+
+	requestType, err := m.submitRequest(req, pb1.NamedCacheRequestType_QueryEntries)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.getStreamingResponseKeyAndValue(ctx, requestType, req.Id)
+}
+
+func (m *streamManagerV1) keySetFilter(ctx context.Context, cache string, filter []byte) (<-chan BinaryKey, error) {
+	req, err := m.newKeySetRequest(cache, filter)
+	if err != nil {
+		return nil, err
+	}
+
+	requestType, err := m.submitRequest(req, pb1.NamedCacheRequestType_QueryKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.getStreamingResponseKey(ctx, requestType, req.Id)
+}
+
+func (m *streamManagerV1) valuesFilter(ctx context.Context, cache string, filter []byte) (<-chan BinaryValue, error) {
+	req, err := m.newValuesFilterRequest(cache, filter)
+	if err != nil {
+		return nil, err
+	}
+
+	requestType, err := m.submitRequest(req, pb1.NamedCacheRequestType_QueryValues)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.getStreamingResponseValue(ctx, requestType, req.Id)
+}
+
+func (m *streamManagerV1) getStreamingResponseKeyAndValue(ctx context.Context, requestType namedCacheRequest, reqID int64, paged ...bool) (<-chan BinaryKeyAndValue, error) {
+	isPaged := false
+	if len(paged) == 1 {
+		// if paged is true the first response will always be the cookie
+		isPaged = paged[0]
+	}
+
+	// create a channel that will be populated by the streaming request
+	ch := make(chan BinaryKeyAndValue)
+
+	newCtx, cancel := m.session.ensureContext(ctx)
+
+	// this channel will receive entries back from the stream
+	respChannel := waitForStreamingResponse(newCtx, requestType.ch)
+
+	go func() {
```

Comment:
This loop is duplicated several times.  Could it be generalized such that you can pass function taking the in coming message and the returning the unpacked response?

Reply:
As discussed, the de-duplication of this is quite complex as they return different types and the initial stab at the code makes it much more complex to maintain.